distributed/safetensor_utils.py

```diff
@@ -0,0 +1,325 @@
+
+            update_state_dict(
+                stage_state_dict,
+                checkpoint,
+                weight_map,
+                new_to_old_keymap,
+                file,
+                updated_states,
+                device,
+            )
+        except FileNotFoundError:
+            logger.error(f"File not found: {full_path}")
+        except Exception as e:
+            logger.error(f"Error during checkpoint processing of {full_path}: {str(e)}")
+
+    missing_keys = handle_missing_keys(
+        stage_state_dict, updated_states, ignore_cache_layers
+    )
+    log_loading_status(missing_keys, updated_states)
+    logger.info(f"Loading {len(updated_states)} weights into stage module")
+    #precount, premap = record_module_dtypes(stage_module)
+    stage_module.load_state_dict(stage_state_dict, strict=False, assign=True)
+    #postcount, postmap = record_module_dtypes(stage_module)
+    #logger.info(f"{precount=}, {postcount=}")
+    #logger.info(f"{premap=}, {postmap=}")
+    #assert False, "checkme"
+
+    return len(updated_states), len(missing_keys)
+
+
+def prepare_state_dict(
+    module: Module, weight_map: Dict[str, str], purge_model_prefix: bool
+) -> Dict[str, torch.Tensor]:
+    state_dict = module.state_dict()
+    if purge_model_prefix:
+        state_dict = {k.removeprefix("model."): v for k, v in state_dict.items()}
+        weight_map = {k.removeprefix("model."): v for k, v in weight_map.items()}
+    return state_dict, weight_map
+
+
+def get_needed_files(
+    state_dict: Dict[str, torch.Tensor], weight_map: Dict[str, str]
+) -> Set[str]:
+    needed_files = set()
+    for param in state_dict.keys():
+        file = weight_map.get(param)
+        if file:
+            needed_files.add(file)
+        elif param.endswith("weight"):
+            logger.warning(
+                f"Parameter {param} not found in weight map, please check..."
+            )
+            raise ValueError(f"Missing file for {param} in {weight_map.keys()}")
+    logger.info(f"Needed files: {needed_files}")
+    return needed_files
+
+
+def load_checkpoint(full_path: str, device: torch.device) -> Dict[str, torch.Tensor]:
+    tensors = {}
+    with safe_open(full_path, framework="pt", device=device) as f:
+        for k in f.keys():
+            tensors[k] = f.get_tensor(k)
+    logger.info(f"Loaded {len(tensors)} tensors from {full_path}")
+    return tensors
+
+
+def update_state_dict(
+    state_dict: Dict[str, torch.Tensor],
+    checkpoint: Dict[str, torch.Tensor],
+    weight_map: Dict[str, str],
+    new_to_old_keymap: Dict[str, str],
+    file: str,
+    updated_states: Set[str],
+    device: torch.device,
+):  
+    count_dtensors_loaded = 0
+    for param, file_with_param in weight_map.items():
+        if file_with_param == file and param in state_dict:
+            model_param = (
+                "output.weight" if param == "output.weight" else f"model.{param}"
+            )
```

Comment:
nit: is this part specific ?

Reply:
we need this otherwise:
WARNING - Partially updated state dict. Missing 1 keys: {'output.weight'} 
if we don't have it.